test/contracts/replayprotection.test.ts

```diff
@@ -0,0 +1,636 @@
+        createReplayProtection
+      );
+
+      for (let queue = 0; queue < 20; queue++) {
+        for (let queueNonce = 0; queueNonce < 20; queueNonce++) {
+          const nonceReplayProtection = defaultAbiCoder.encode(
+            ["uint", "uint"],
+            [queue, queueNonce]
+          );
+
+          await replayProtection.noncePublic(
+            admin.address,
+            nonceReplayProtection
+          );
+
+          const index = keccak256(
+            defaultAbiCoder.encode(
+              ["address", "uint", "address"],
+              [admin.address, queue, AddressZero]
+            )
+          );
+          const storedNonce = await replayProtection.nonceStore(index);
+
+          expect(storedNonce.toNumber()).to.eq(queueNonce + 1);
+        }
+      }
+    }
+  ).timeout("50000");
+
+  fnIt<replayProtection>(
+    (a) => a.verifyPublic,
+    "access nonce() via verify (e.g. the replay protection authority)",
+    async () => {
+      const { replayProtection, admin } = await loadFixture(
+        createReplayProtection
+      );
+
+      const { callData, encodedReplayProtection, signedCall } = await signCall(
+        replayProtection,
+        AddressZero,
+        admin,
+        0,
+        0
+      );
+
+      const index = keccak256(
+        defaultAbiCoder.encode(
+          ["address", "uint", "address"],
+          [admin.address, 0, AddressZero]
+        )
+      );
+
+      await replayProtection.verifyPublic(
+        callData,
+        encodedReplayProtection,
+        AddressZero,
+        admin.address,
+        signedCall
+      );
+
+      const nonce = await replayProtection.nonceStore(index);
+
+      expect(nonce).to.eq(1);
+    }
+  ).timeout("50000");
+
+  fnIt<replayProtection>(
+    (a) => a.verifyPublic,
+    "for nonce() the same replay protection is used twice and should fail",
+    async () => {
+      const { replayProtection, admin } = await loadFixture(
+        createReplayProtection
+      );
+
+      const { callData, encodedReplayProtection, signedCall } = await signCall(
+        replayProtection,
+        AddressZero,
+        admin,
+        0,
+        0
+      );
+
+      const index = keccak256(
+        defaultAbiCoder.encode(
+          ["address", "uint", "address"],
+          [admin.address, 0, "0x0000000000000000000000000000000000000000"]
+        )
+      );
+
+      await replayProtection.verifyPublic(
+        callData,
+        encodedReplayProtection,
+        AddressZero,
+        admin.address,
+        signedCall
+      );
+
+      const nonce = await replayProtection.nonceStore(index);
+
+      expect(nonce).to.eq(1);
+
+      const tx = replayProtection.verifyPublic(
+        callData,
+        encodedReplayProtection,
+        AddressZero,
+        admin.address,
+        signedCall
+      );
+
+      await expect(tx).to.be.revertedWith(
+        "Multinonce replay protection failed"
+      );
+    }
+  ).timeout("50000");
```

Comment:
shouldn't have a large timeout for tests that don't need it (code taking too long unexpectedly is a possible bug that unit tests could catch).

Reply:
i removed some of them. kept some for the endToEnd, since they do take a long time.